src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,5246 @@
+        {
+            _controlFlags = (byte)(controlFlags & ControlMask);
+            _tagValue = tagValue;
+        }
+
+        public Asn1Tag(UniversalTagNumber universalTagNumber, bool isConstructed = false)
+            : this(isConstructed ? ConstructedMask : (byte)0, (int)universalTagNumber)
+        {
+            const UniversalTagNumber ReservedIndex = (UniversalTagNumber)15;
+
+            if (universalTagNumber < UniversalTagNumber.EndOfContents ||
+                universalTagNumber > UniversalTagNumber.Duration ||
+                universalTagNumber == ReservedIndex)
+            {
+                throw new ArgumentOutOfRangeException(nameof(universalTagNumber), universalTagNumber, null);
+            }
+        }
+
+        public Asn1Tag(TagClass tagClass, int tagValue, bool isConstructed = false)
+            : this((byte)((byte)tagClass | (isConstructed ? ConstructedMask : 0)), tagValue)
+        {
+            if (tagClass < TagClass.Universal || tagClass > TagClass.Private)
+            {
+                throw new ArgumentOutOfRangeException(nameof(tagClass), tagClass, null);
+            }
+        }
+
+        public Asn1Tag(byte singleByteEncoding)
+        {
+            Span<byte> span = stackalloc byte[1];
+            span[0] = singleByteEncoding;
+
+            if (TryParse(span, out Asn1Tag parsed, out int bytesRead))
+            {
+                Debug.Assert(bytesRead == 1);
+                _controlFlags = parsed._controlFlags;
+                _tagValue = parsed._tagValue;
+            }
+            else
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+        }
+
+        public static bool TryParse(ReadOnlySpan<byte> source, out Asn1Tag tag, out int bytesRead)
+        {
+            tag = default(Asn1Tag);
+            bytesRead = 0;
+
+            if (source.IsEmpty)
+                return false;
+
+            byte first = source[bytesRead];
+            bytesRead++;
+            uint tagValue = (uint)(first & TagNumberMask);
+
+            if (tagValue == TagNumberMask)
+            {
+                // Multi-byte encoding
+                // T-REC-X.690-201508 sec 8.1.2.4
+                const byte ContinuationFlag = 0x80;
+                const byte ValueMask = ContinuationFlag - 1;
+
+                tagValue = 0;
+                byte current;
+
+                do
+                {
+                    if (source.Length <= bytesRead)
+                    {
+                        bytesRead = 0;
+                        return false;
+                    }
+
+                    current = source[bytesRead];
+                    byte currentValue = (byte)(current & ValueMask);
+                    bytesRead++;
+
+                    // The first byte cannot have the value 0 (T-REC-X.690-201508 sec 8.1.2.4.2.c)
+                    if (currentValue == 0 && tagValue == 0)
```

Comment:
Nit: move this check to the end of the loop (just before the _while_) and you can omit the `currentValue == 0` clause.